### PR TITLE
add freeze feature to rundown

### PR DESCRIPTION
**Pull request by jwetzell**

### CODE REVIEW

File: apps/server/src/app.ts [hunk 1]
```diff
@@ -188,6 +188,7 @@ export const startServer = async (
       playback: SimplePlayback.Stop,
       direction: SimpleDirection.CountDown,
     },
+    frozen: false,
```
[the remark this answers is not in the document]
Comment (jwetzell) [reply apps/server/src/app.ts:191]: LOL what?

[the remark this answers is not in the document]
Comment (jwetzell) [reply apps/server/src/app.ts:191]: Silly rabbit that variable has not been declared yet. `startServer` doesn't have a parameter called `frozen`.

[the remark this answers is not in the document]
Comment (jwetzell) [reply apps/server/src/app.ts:191]: no thanks

File: apps/server/src/api-data/rundown/rundown.controller.ts [hunk 2]
```diff
@@ -69,6 +70,21 @@ export async function rundownBatchPut(req: Request, res: Response<MessageRespons
   }
 }
 
+export async function rundownFrozenPost(req: Request, res: Response<MessageResponse | ErrorResponse>) {
+  if (failEmptyObjects(req.body, res)) {
+    return res.status(404);
+  }
```
Comment (cpvalente) [inline apps/server/src/api-data/rundown/rundown.controller.ts:76]: question: would this not have been taken care of by the validator?

File: apps/server/src/api-data/rundown/rundown.controller.ts [hunk 3]
```diff
@@ -69,6 +70,21 @@ export async function rundownBatchPut(req: Request, res: Response<MessageRespons
...
+
+  try {
+    const { state } = req.body;
```
Comment (cpvalente) [inline apps/server/src/api-data/rundown/rundown.controller.ts:79]: question: could this name be more descriptive? even if it is just `frozen`?

File: apps/server/src/api-data/rundown/rundown.controller.ts [hunk 4]
```diff
@@ -69,6 +70,21 @@ export async function rundownBatchPut(req: Request, res: Response<MessageRespons
...
+    setFrozenState(state);
+    res.status(200).send({ message: `Rundown frozen state updated.` });
```
Comment (cpvalente) [inline apps/server/src/api-data/rundown/rundown.controller.ts:81]: suggestion: no need for template literals here

**Review by cpvalente — COMMENTED** [review]
Code looks good, I made some suggestions that you can consider

I will apply the frontend part and see how the feature behaves.
Thank you for your help

File: apps/server/src/api-data/rundown/rundown.controller.ts (hunk 2 above)
Comment (jwetzell) [reply apps/server/src/api-data/rundown/rundown.controller.ts:76]: Most of the other functions here had that check at the top so I just followed suit.

Comment (cpvalente) [reply apps/server/src/api-data/rundown/rundown.controller.ts:76]: ah I see.
There is a case where the other code is also wrong,

But in some cases, all the elements of the body are optional, so the validation could potentially pass with an empty object

Comment (jwetzell) [reply apps/server/src/api-data/rundown/rundown.controller.ts]: Fair. I've cleaned that out and this should be good to go now.

**Comment by cpvalente** [comment]
Thank you @jwetzell , I will continue the PR with the client implementation

**Pull request merged**